crates/noirc_evaluator/src/ssa_refactor/cfg.rs

```diff
@@ -0,0 +1,110 @@
+
+    /// Set of blocks that are the targets of jumps in this block.
+    /// The successors set has no meaningful order.
+    pub(crate) successors: HashSet<BasicBlockId>,
+}
+
+/// The Control Flow Graph maintains a mapping of blocks to their predecessors
+/// and successors where predecessors are basic blocks and successors are
+/// basic blocks.
+pub(crate) struct ControlFlowGraph {
+    data: HashMap<BasicBlockId, CfgNode>,
+}
+
+impl ControlFlowGraph {
+    /// Allocate and compute the control flow graph for `func`.
+    pub(crate) fn with_function(func: &Function) -> Self {
+        let mut cfg = ControlFlowGraph { data: HashMap::new() };
+        cfg.compute(func);
+        cfg
+    }
+
+    fn compute(&mut self, func: &Function) {
+        for basic_block_id in func.basic_block_ids_iter() {
+            self.compute_block(func, basic_block_id);
+        }
+    }
+
+    fn compute_block(&mut self, func: &Function, basic_block_id: BasicBlockId) {
+        instruction_predicates::visit_block_succs(func, basic_block_id, |dest| {
+            self.add_edge(basic_block_id, dest);
+        });
+    }
+
+    fn invalidate_block_successors(&mut self, basic_block_id: BasicBlockId) {
+        let node = self
+            .data
+            .get_mut(&basic_block_id)
+            .expect("ICE: Attempted to invalidate cfg node successors for non-existent node.");
+        let old_successors = node.successors.clone();
+        node.successors.clear();
+        for successor_id in old_successors {
+            self.data
+                .get_mut(&successor_id)
+                .expect("ICE: Cfg node successor doesn't exist.")
+                .predecessors
+                .remove(&basic_block_id);
+        }
+    }
+
+    /// Recompute the control flow graph of `block`.
+    ///
+    /// This is for use after modifying instructions within a specific block. It recomputes all edges
+    /// from `basic_block_id` while leaving edges to `basic_block_id` intact.
+    pub(crate) fn recompute_block(&mut self, func: &Function, basic_block_id: BasicBlockId) {
+        self.invalidate_block_successors(basic_block_id);
+        self.compute_block(func, basic_block_id);
+    }
+
+    fn add_edge(&mut self, from: BasicBlockId, to: BasicBlockId) {
+        let predecessor_node = self.data.entry(from).or_default();
+        assert!(
+            predecessor_node.successors.len() < 2,
+            "ICE: A cfg node cannot have more than two successors"
+        );
+        predecessor_node.successors.insert(to);
+        let successor_node = self.data.entry(to).or_default();
+        assert!(
+            successor_node.predecessors.len() < 2,
+            "ICE: A cfg node cannot have more than two predecessors"
+        );
+        successor_node.predecessors.insert(from);
+    }
+
+    /// Get an iterator over the CFG predecessors to `block`.
+    pub(crate) fn pred_iter(&self, basic_block_id: BasicBlockId) -> PredIter {
```

Comment:
Instead of PredIter & SuccIter aliases, can we use `impl Iterator<Item = BasicBlockId>` (or ExactSizeIter) for both?